### PR TITLE
Prevent object parent recursion

### PR DESCRIPTION
**Pull request by denisvmedyantsev**

### CODE REVIEW

File: savant/deepstream/buffer_processor.py [hunk 1]
```diff
@@ -606,7 +606,9 @@ def prepare_element_output(self, element: PipelineElement, buffer: Gst.Buffer):
                         # for object/complex models output - `bbox_tensor` and
                         # `selected_bboxes` - indices of selected bboxes and meta
                         # for attribute/complex models output - `values`
-                        bbox_tensor, selected_bboxes, values = None, None, None
+                        bbox_tensor: Optional[np.array] = None
+                        selected_bboxes: Optional[list] = None
```
Comment (dorgun) [inline savant/deepstream/buffer_processor.py]:
```suggestion
                        selected_bboxes: Optional[List] = None
```
Since we're still using python 3.8, it's better to use typing generics

File: savant/deepstream/buffer_processor.py [hunk 2]
```diff
@@ -606,7 +606,9 @@ def prepare_element_output(self, element: PipelineElement, buffer: Gst.Buffer):
                         # for object/complex models output - `bbox_tensor` and
                         # `selected_bboxes` - indices of selected bboxes and meta
                         # for attribute/complex models output - `values`
-                        bbox_tensor, selected_bboxes, values = None, None, None
+                        bbox_tensor: Optional[np.array] = None
```
Comment (dorgun) [inline savant/deepstream/buffer_processor.py]:
```suggestion
                        bbox_tensor: Optional[np.ndarray] = None
```

File: savant/deepstream/buffer_processor.py [hunk 3]
```diff
@@ -606,7 +606,9 @@ def prepare_element_output(self, element: PipelineElement, buffer: Gst.Buffer):
...
+                        selected_bboxes: Optional[list] = None
+                        values: Optional[list] = None
```
Comment (dorgun) [inline savant/deepstream/buffer_processor.py]:
```suggestion
                        values: Optional[List] = None
```

File: savant/deepstream/meta/object.py [hunk 4]
```diff
@@ -75,20 +75,11 @@ def __init__(
...
-                parent.object_meta_impl.ds_object_meta if parent is not None else None,
-            )
-        else:
-            raise MetaValueError(
-                f'parent is `{type(parent)}` type, expected `{self.__class__.__name__}`'
-            )
+        pyds.nvds_add_obj_meta_to_frame(
+            self._frame_meta.frame_meta,
+            self.ds_object_meta,
+            parent.object_meta_impl.ds_object_meta if parent is not None else None,
```
Comment (dorgun) [inline savant/deepstream/meta/object.py:81]: Why did you remove the check? The user can simply create an ObjectMeta and specify it as a parent. When an ObjectMeta is created, it is not immediately added to the DS meta and therefore its object_meta_impl will be None. Eventually object_meta_impl will be None and there will be an error.

File: savant/deepstream/meta/object.py [hunk 5]
```diff
@@ -295,21 +289,25 @@ def element_name(self) -> str:
...
         :param frame_meta: Deepstream frame meta.
+        :param depth: Object parent recursion depth.
         :return:
         """
         self = cls.__new__(cls)
         self.ds_object_meta = object_meta
         self._frame_meta = frame_meta
         self._bbox = None
-        if not nvds_is_empty_object_meta(object_meta.parent):
+        if not nvds_is_empty_object_meta(object_meta.parent) and depth > 0:
```
Comment (dorgun) [inline savant/deepstream/meta/object.py:308]: My suggestion is to just move this code to getter parent and make it lazy to create a wrapper for parent. Then we won't need to check the recursion depth. Or at least give some warning here. But now it is just none when it is deeper than 2 and that's all. What if a user actually creates a relation with a depth greater than 2?

**Review by dorgun — CHANGES_REQUESTED** [review]

File: savant/deepstream/meta/object.py (hunk 5 above)
Comment (bwsw) [reply savant/deepstream/meta/object.py:308]: @denisvmedyantsev @dorgun ignore that, yet. Not worth doing extra engineering.

File: savant/deepstream/meta/object.py (hunk 4 above)
Comment (denisvmedyantsev) [reply savant/deepstream/meta/object.py:81]: Because a line above `self.parent = parent` (parent setter) checks the same.

Comment (denisvmedyantsev) [reply savant/deepstream/meta/object.py:81]: Because a line above `self.parent = parent` (parent setter) checks the same.

File: savant/deepstream/buffer_processor.py (hunk 1 above)
Comment (denisvmedyantsev) [reply savant/deepstream/buffer_processor.py]: done

File: savant/deepstream/buffer_processor.py (hunk 2 above)
Comment (denisvmedyantsev) [reply savant/deepstream/buffer_processor.py]: done

File: savant/deepstream/buffer_processor.py (hunk 3 above)
Comment (denisvmedyantsev) [reply savant/deepstream/buffer_processor.py]: done

**Review by dorgun — APPROVED** [review]
